Assets/Scripts/FpsController.cs

```diff
@@ -0,0 +1,158 @@
+    public Camera cam;
+    public float lookHorizontalSensitivity = 1.0f;
+    public float lookVerticalSensitivity = 1.0f;
+    public float minimumPitch = -89.9f;
+    public float maximumPitch = 89.9f;
+
+    [Header("Jump Settings")]
+    public float jumpForce = 300.0f;
+    public float jumpInputBufferTime = 0.25f;
+
+    [Header("Movement Settings")]
+    [Range(0.1f, 15.0f)]
+    public float forwardSpeed = 5.0f;
+    [Range(0.1f, 15.0f)]
+    public float backSpeed = 3.0f;
+
+    private Vector2 _inputVector = Vector2.zero;
+    private bool _jumpInput = false;
+    private float _jumpTimeSinceInput = 0.0f;
+    private bool _grounded = false;
+
+    private float _lookYaw;
+    private float _lookPitch;
+
+    void Reset()
+    {
+        body = GetComponent<Rigidbody>();
+        cam = GetComponentInChildren<Camera>();
+    }
+
+    void Start()
+    {
+        Cursor.lockState = CursorLockMode.Locked;
+        _lookYaw = transform.eulerAngles.y;
+        _lookPitch = cam.transform.eulerAngles.x;
+    }
+
+    void Update()
+    {
+        Look();
+        MovementInput();
+    }
+
+    void FixedUpdate()
+    {
+        MoveLogic();
+        JumpLogic();
+
+        UpdateGroundedState();
+    }
+
+    void ApplyDirectionalSpeedMultipliers(ref Vector3 vector)
+    {
+        if (_inputVector.y < 0.0f)
+        {
+            vector = vector.normalized * backSpeed;
+        }
+        else
+        {
+            vector = vector.normalized * forwardSpeed;
+        }
+    }
+
+    Vector3 GetMovementVector()
+    {
+        return transform.TransformVector(new Vector3(_inputVector.x, 0.0f, _inputVector.y));
+    }
+
+    void Move(Vector3 displacement)
+    {
+        transform.position += displacement;
+    }
+
+    void Look()
+    {
+        float mouseDeltaYaw = Input.GetAxisRaw("Mouse X");
+        float mouseDeltaPitch = Input.GetAxisRaw("Mouse Y");
+
+        _lookYaw += mouseDeltaYaw;
+        _lookYaw = WrapAroundAngleDegrees(_lookYaw);
+        transform.rotation = Quaternion.Euler(transform.eulerAngles.x, _lookYaw, transform.eulerAngles.z);
+
+        _lookPitch -= mouseDeltaPitch;
+        _lookPitch = ClampPitch(_lookPitch);
+        cam.transform.rotation = Quaternion.Euler(_lookPitch, cam.transform.eulerAngles.y, cam.transform.eulerAngles.z);
+    }
+
+    void MovementInput()
+    {
+        _inputVector = new Vector2(Input.GetAxis("Horizontal"), Input.GetAxis("Vertical"));
+
+        if (_jumpTimeSinceInput >= jumpInputBufferTime)
+        {
+            _jumpInput = Input.GetButtonDown("Jump");
+            if (_jumpInput)
+            {
+                _jumpTimeSinceInput = 0.0f;
+            }
+        }
+        else
+        {
+            _jumpTimeSinceInput += Time.deltaTime;
+        }
+    }
+
+    void UpdateGroundedState()
+    {
+        if (Physics.Raycast(transform.position, Vector3.down, out RaycastHit hitInfo)
+        && body.velocity.y < 0.01f
+        && hitInfo.distance <= 2.0f / 2.0f)
```

Comment:
Nice 2.0f / 2.0f :+1:

As the `Vector3.down` is only 1 in length, it's already only checking a max distance of 1, so this can be shortened as:

```suggestion
        if (body.velocity.y < 0.01f
            && Physics.Raycast(transform.position, Vector3.down, out RaycastHit hitInfo))
```